File: crates/agentpin-cli/src/verify.rs
use std::fs;

use clap::Args;

use agentpin::pinning::KeyPinStore;
use agentpin::resolver::{
    ChainResolver, DiscoveryResolver, LocalFileResolver, TrustBundleResolver,
};
use agentpin::types::bundle::TrustBundle;
use agentpin::types::discovery::DiscoveryDocument;
use agentpin::types::revocation::RevocationDocument;
use agentpin::verification::{
    verify_credential, verify_credential_offline, verify_credential_with_resolver, VerifierConfig,
};

#[derive(Args)]
pub struct VerifyArgs {
    /// JWT credential string or path to file containing it
    #[arg(long)]
    pub credential: String,

    /// Path to discovery document JSON file (for offline verification)
    #[arg(long)]
    pub discovery: Option<String>,

    /// Path to revocation document JSON file (for offline verification)
    #[arg(long)]
    pub revocation: Option<String>,

    /// Path to pin store JSON file
    #[arg(long)]
    pub pin_store: Option<String>,

    /// Verifier's audience domain
    #[arg(long)]
    pub audience: Option<String>,

    /// Use offline-only verification (no HTTP fetches)
    #[arg(long)]
    pub offline: bool,

    /// Path to a trust bundle JSON file for verification
    #[arg(long)]
    pub trust_bundle: Option<String>,

    /// Path to a directory containing discovery document files ({domain}.json)
    #[arg(long)]
    pub discovery_dir: Option<String>,
}

pub async fn run(args: VerifyArgs) -> anyhow::Result<()> {
    // Load credential — could be a JWT string or a file path
    let credential = if std::path::Path::new(&args.credential).exists() {
        fs::read_to_string(&args.credential)?.trim().to_string()
    } else {
        args.credential.clone()
    };

    // Load or create pin store
    let mut pin_store = match &args.pin_store {
        Some(path) if std::path::Path::new(path).exists() => {
            let json = fs::read_to_string(path)?;
            KeyPinStore::from_json(&json)?
        }
        _ => KeyPinStore::new(),
    };

    let config = VerifierConfig::default();

    let result = if args.trust_bundle.is_some() || args.discovery_dir.is_some() {
        // Resolver-based verification
        let resolver = build_resolver(&args)?;
        verify_credential_with_resolver(
            &credential,
            resolver.as_ref(),
            &mut pin_store,
            args.audience.as_deref(),
            &config,
        )
    } else if args.offline || args.discovery.is_some() {
        // Offline verification
        let discovery_path = args
            .discovery
            .as_ref()
            .ok_or_else(|| anyhow::anyhow!("--discovery is required for offline verification"))?;
        let discovery_json = fs::read_to_string(discovery_path)?;
        let discovery: DiscoveryDocument = serde_json::from_str(&discovery_json)?;

        let revocation: Option<RevocationDocument> = match &args.revocation {
            Some(path) => {
                let json = fs::read_to_string(path)?;
                Some(serde_json::from_str(&json)?)
            }
            None => None,
        };

        verify_credential_offline(
            &credential,
            &discovery,
            revocation.as_ref(),
            &mut pin_store,
            args.audience.as_deref(),
            &config,
        )
    } else {
        // Online verification — fetch discovery/revocation from issuer domain
        verify_credential(
            &credential,
            &mut pin_store,
            args.audience.as_deref(),
            &config,
        )
        .await
    };

    let output = serde_json::to_string_pretty(&result)?;
    println!("{}", output);

    // Persist pin store if path was provided
    if let Some(path) = &args.pin_store {
        let json = pin_store.to_json()?;
        fs::write(path, json)?;
    }

    if !result.valid {
        std::process::exit(1);
    }

    Ok(())
}

/// Build a sync resolver from the CLI flags.
fn build_resolver(args: &VerifyArgs) -> anyhow::Result<Box<dyn DiscoveryResolver>> {
    let mut resolvers: Vec<Box<dyn DiscoveryResolver>> = Vec::new();

    if let Some(ref path) = args.trust_bundle {
        let json = fs::read_to_string(path)?;
        let bundle: TrustBundle = serde_json::from_str(&json)?;
        resolvers.push(Box::new(TrustBundleResolver::new(&bundle)));
    }

    if let Some(ref dir) = args.discovery_dir {
        resolvers.push(Box::new(LocalFileResolver::new(
            std::path::Path::new(dir),
            None,
        )));
    }

    match resolvers.len() {
        0 => Err(anyhow::anyhow!(
            "At least --trust-bundle or --discovery-dir must be provided"
        )),
        1 => Ok(resolvers.remove(0)),
        _ => Ok(Box::new(ChainResolver::new(resolvers))),
    }
}

File: go/cmd/agentpin/verify.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"

	"github.com/ThirdKeyAi/agentpin/go/pkg/pinning"
	"github.com/ThirdKeyAi/agentpin/go/pkg/resolver"
	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
	"github.com/ThirdKeyAi/agentpin/go/pkg/verification"
)

func runVerify(args []string) error {
	fs := flag.NewFlagSet("verify", flag.ContinueOnError)
	credentialArg := fs.String("credential", "", "JWT credential string or file path (required)")
	discoveryPath := fs.String("discovery", "", "Path to discovery document JSON file (offline)")
	revocationPath := fs.String("revocation", "", "Path to revocation document JSON file (offline)")
	pinStorePath := fs.String("pin-store", "", "Path to pin store JSON file")
	audience := fs.String("audience", "", "Verifier's audience domain")
	offline := fs.Bool("offline", false, "Use offline-only verification")
	trustBundlePath := fs.String("trust-bundle", "", "Trust bundle JSON file for resolver mode")
	discoveryDir := fs.String("discovery-dir", "", "Directory of {domain}.json discovery files for resolver mode")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *credentialArg == "" {
		return fmt.Errorf("--credential is required")
	}

	cred := *credentialArg
	if _, err := os.Stat(cred); err == nil {
		data, err := os.ReadFile(cred)
		if err != nil {
			return err
		}
		cred = strings.TrimSpace(string(data))
	}

	pinStore := pinning.NewKeyPinStore()
	if *pinStorePath != "" {
		if data, err := os.ReadFile(*pinStorePath); err == nil {
			if err := pinStore.LoadFromJSON(data); err != nil {
				return fmt.Errorf("load pin store: %w", err)
			}
		}
	}
	cfg := verification.DefaultVerifierConfig()

	var result verification.Result
	if *trustBundlePath != "" || *discoveryDir != "" {
		r, err := buildResolver(*trustBundlePath, *discoveryDir)
		if err != nil {
			return err
		}
		result = verification.VerifyCredentialWithResolver(cred, r, pinStore, *audience, cfg)
	} else if *offline || *discoveryPath != "" {
		if *discoveryPath == "" {
			return fmt.Errorf("--discovery is required for offline verification")
		}
		dData, err := os.ReadFile(*discoveryPath)
		if err != nil {
			return err
		}
		var disc types.DiscoveryDocument
		if err := json.Unmarshal(dData, &disc); err != nil {
			return err
		}
		var rev *types.RevocationDocument
		if *revocationPath != "" {
			rData, err := os.ReadFile(*revocationPath)
			if err != nil {
				return err
			}
			var r types.RevocationDocument
			if err := json.Unmarshal(rData, &r); err != nil {
				return err
			}
			rev = &r
		}
		result = verification.VerifyCredentialOffline(cred, &disc, rev, pinStore, *audience, cfg)
	} else {
		// Online verification via WellKnownResolver.
		r := resolver.NewWellKnownResolver()
		result = verification.VerifyCredentialWithResolver(cred, r, pinStore, *audience, cfg)
	}

	out, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return err
	}
	fmt.Println(string(out))

	if *pinStorePath != "" {
		data, err := pinStore.MarshalJSON()
		if err == nil {
			_ = os.WriteFile(*pinStorePath, data, 0o644)
		}
	}

	if !result.Valid {
		os.Exit(1)
	}
	return nil
}

func buildResolver(trustBundlePath, discoveryDir string) (resolver.DiscoveryResolver, error) {
	resolvers := []resolver.DiscoveryResolver{}
	if trustBundlePath != "" {
		data, err := os.ReadFile(trustBundlePath)
		if err != nil {
			return nil, err
		}
		r, err := resolver.TrustBundleResolverFromJSON(data)
		if err != nil {
			return nil, err
		}
		resolvers = append(resolvers, r)
	}
	if discoveryDir != "" {
		resolvers = append(resolvers, resolver.NewLocalFileResolver(discoveryDir, ""))
	}
	if len(resolvers) == 0 {
		return nil, fmt.Errorf("at least --trust-bundle or --discovery-dir must be provided")
	}
	if len(resolvers) == 1 {
		return resolvers[0], nil
	}
	return resolver.NewChainResolver(resolvers), nil
}
